### ISSUE TITLE
Tag overwrite does not work correctly

### ISSUE THREAD

**Issue by Lopuiz** [wazuh/wazuh#3807]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.9.3 | Analysisd-Ruleset | Manager | Packages/Sources | All |

---
Hello team,

I'm working on Ruleset tests and I need to overwrite the following rule:
```xml
  <rule id="60000" level="0">
    <category>ossec</category>
    <decoded_as>windows_eventchannel</decoded_as>
    <field name="win.system.providerName">\.+</field>
    <options>no_full_log</options>
    <description>Group of windows rules</description>
  </rule>
```
For this one:
```xml
  <rule id="60000" level="0" overwrite="yes">
    <decoded_as>json</decoded_as>
    <field name="win.system.providerName">\.+</field>
    <options>no_full_log</options>
    <description>Overwrite succefuly</description>
  </rule>
```
But this configuration doesn't work. The problem is that the following function does not overwrite the category tag when overwriting a rule:
https://github.com/wazuh/wazuh/blob/852146faece1c041c8b90045d8a53bd6a4f037a1/src/analysisd/rules_list.c#L256-L308

Regards,
Eva

**Comment by Lopuiz** [comment]
| Working branch |
| ---- |
| [3.10-fix-overwriteCategory-3807](https://github.com/wazuh/wazuh/tree/3.10-fix-overwriteCategory-3807) |

**Comment by Lopuiz** [comment]
Hi!

I've found a new bug. Tag `if_sid` is also no overwritten.
This value is not modified in the following function:
https://github.com/wazuh/wazuh/blob/852146faece1c041c8b90045d8a53bd6a4f037a1/src/analysisd/rules_list.c#L256

---
To check it, you can use the following rules:
```xml
<group name="testing">

  <!--
  Dec 10 01:02:02 my_host my_progrma[1234]: ifsid test
  -->
   <rule id="999999" level="1">
      <match>ifsid test</match>
      <description>Parent rule</description>
   </rule>

   <rule id="999998" level="4">
      <if_sid>999999</if_sid>
      <match>r98</match>
      <description>Child rule 98</description>
   </rule>

   <rule id="999997" level="4">
      <if_sid>999999</if_sid>
      <match>r97</match>
      <description>Child rule 97</description>
   </rule>

      <rule id="999996" level="6">
      <if_sid>999998</if_sid>
      <match>r96</match>
      <description>Child rule 98 96</description>
   </rule>
   <rule id="999996" level="4" overwrite="yes">
      <if_sid>999997</if_sid>
      <match>r96</match>
      <description>Child overwrite rule 97 96</description>
   </rule>

</group>
```
Ossec-logtest output:
```
root@lopezziur:/var/ossec# bin/ossec-logtest
2019/08/08 08:54:58 ossec-testrule: INFO: Started (pid: 11488).
ossec-testrule: Type one log per line.

Dec 10 01:02:02 my_host my_progrma[1234]: ifsid test r97 r96


**Phase 1: Completed pre-decoding.
       full event: 'Dec 10 01:02:02 my_host my_progrma[1234]: ifsid test r97 r96'
       timestamp: 'Dec 10 01:02:02'
       hostname: 'my_host'
       program_name: 'my_progrma'
       log: 'ifsid test r97 r96'

**Phase 2: Completed decoding.
       No decoder matched.

**Phase 3: Completed filtering (rules).
       Rule id: '999997'
       Level: '4'
       Description: 'Child rule 97'
**Alert to be generated.


Dec 10 01:02:02 my_host my_progrma[1234]: ifsid test r98 r96


**Phase 1: Completed pre-decoding.
       full event: 'Dec 10 01:02:02 my_host my_progrma[1234]: ifsid test r98 r96'
       timestamp: 'Dec 10 01:02:02'
       hostname: 'my_host'
       program_name: 'my_progrma'
       log: 'ifsid test r98 r96'

**Phase 2: Completed decoding.
       No decoder matched.

**Phase 3: Completed filtering (rules).
       Rule id: '999996'
       Level: '4'
       Description: 'Child overwrite rule 97 96'
**Alert to be generated.
```

**Comment by Lopuiz** [comment]
Hello team!

I'm sorry the above comments don't describe the problem well. I will try to describe the problem with the following diagram.
This is an example of how the rules are saved in Wazuh:

![Untitled Diagram](https://user-images.githubusercontent.com/17710550/62700196-98920f80-b9e1-11e9-9f62-c3a0554638e6.png)

The nodes structure is as follows:
```c
typedef struct _RuleNode {
    RuleInfo *ruleinfo;
    struct _RuleNode *next;
    struct _RuleNode *child;
} RuleNode;
```
where the `RuleInfo` structure contains all the information related to the rule.

The `OS_AddRuleInfo` function does not overwrite the `if_sid` and `category` tags. And it is the problem to solve. To do this I have:
* assigned the new values
* modified the structure so that the the new parent contains this rule in list `child`.

Regards,
Eva

**Comment by Lopuiz** [comment]
### Update:
1. **if_matched_sid and if_matched_group**
The labels `if_matched_sid` and `if_matched_group` are also not overwritten because not all RuleInfo variables are modified in `OS_AddRuleInfo`. 
For tags to be overwritten the new values of `sid_prev_matched`, `sid_search`, `group_prev_matched`, `group_search`, `event_search` `if_matched_group`, `if_matched_sid` must be assigned.
It's solved in this [commit](https://github.com/wazuh/wazuh/commit/a3c746d1269f4e5ad0c5493de335512df0f11764).
2. **category, if_sid, if_group and if_level**
These tags determine where and how many times a rule will be saved in the structure. To overwrite them, we must modify the structure.
Although I've made modifications to fix this, I don't think it works for all possible use cases. I need to check it again.

### Note:
The `if_level`, `if_matched_sid` and `if_matched_group` tags do not work properly. But they are different issues.
+ [Issue if_matched_group](https://github.com/wazuh/wazuh/issues/3291)


Regards,
Eva

**Comment by mikykeane** [comment]
Hello team, 

We have found an issue that may be related to this one. Overwrite tag is not working as expected with `field` label. 

Using this sample rules:

```
  <rule id="100050" level="5">
    <decoded_as>json</decoded_as>
    <field name="prueba">1</field>
    <description>Probando Json 1</description>
  </rule>

  <rule id="100050" level="6" overwrite="yes">
   <decoded_as>json</decoded_as>
    <field name="prueba">2</field>
    <description>Probando 2</description>
  </rule>
```
If we try this log, which should not trigger an alert we get:


```
{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "1"}


**Phase 1: Completed pre-decoding.
       full event: '{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "1"}'
       timestamp: '(null)'
       hostname: 'master'
       program_name: '(null)'
       log: '{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "1"}'

**Phase 2: Completed decoding.
       decoder: 'json'
       timestamp: '2016-05-02T17:46:48.515262+0000'
       flow_id: '1234'
       prueba: '1'

**Phase 3: Completed filtering (rules).
       Rule id: '100050'
       Level: '6'
       Description: 'Probando 2'

```

The rule has been wrongly triggered by the first rule but changes the Description and Level from the second rule.

Then if we try this other log that should trigger an alert:

```
{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "2"}


**Phase 1: Completed pre-decoding.
       full event: '{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "2"}'
       timestamp: '(null)'
       hostname: 'master'
       program_name: '(null)'
       log: '{"timestamp":"2016-05-02T17:46:48.515262+0000","flow_id": "1234", "prueba": "2"}'

**Phase 2: Completed decoding.
       decoder: 'json'
       timestamp: '2016-05-02T17:46:48.515262+0000'
       flow_id: '1234'
       prueba: '2'

```

No alert is triggered. 

The rule is being overwritten, but it keeps the first label of `field` without overwriting it, giving undesired results. 


Best regards, 
Miguel Keane

**Comment by Canon88** [comment]
comment

**Comment by Canon88** [comment]
Hi Wazuh team:

Unfortunately, I also encountered a problem.  Will you consider fixing this problem in the short term?

**Comment by jnasselle** [comment]
Closed by #8210
